### PR TITLE
Adding device-bootstrap, improving device-test

### PR DESCRIPTION
**Pull request by Ellerbach**
## Description

Adding device-bootstrap, improving device-test

## Motivation and Context

- Adding device-bootstrap as a template for real hardware tests
- Improving device-test

## How Has This Been Tested?

On nanoFramework.Iot.TestStream, See PR https://github.com/nanoframework/nanoFramework.IoT.TestStream/pull/22

## Types of changes

- [x] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [x] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Unit Tests (add new Unit Test(s) or improved existing one(s), has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:

- [x] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [x] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).
- [ ] I have added new tests to cover my changes.

## Summary by CodeRabbit

- **New Features**
	- Introduced a new pipeline configuration for bootstrapping Azure DevOps projects.
	- Enhanced testing stages for both Linux and Windows environments with improved error handling.
  
- **Bug Fixes**
	- Added conditions to check for non-zero test failures in Linux tests to improve reliability.

- **Documentation**
	- Improved clarity and consistency in YAML file formatting and task display names.

### CODE REVIEW

File: azure-pipelines-templates/device-bootstrap.yml [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+      # Define the Azure DevOps organization, project, and pipeline
+      organization="${{ parameters.AZURE_DEVOPS_ORG }}"
+      project="${{ parameters.AZURE_DEVOPS_PROJECT }}"
+      pipelineId="${{ parameters.AZURE_DEVOPS_PIPELINE_ID }}"
+      poolName="${{ parameters.AZURE_POOL_NAME }}"
+      branch="${BUILD_SOURCEBRANCH}"
+
+      # Encode the PAT
+      patEncoded=$(echo -n ":${AZURE_DEVOPS_PAT}" | base64)
```
[the remark this answers is not in the document]
Comment (josesimoes) [reply azure-pipelines-templates/device-bootstrap.yml:35]: This looks like a valid suggestion

File: azure-pipelines-templates/device-bootstrap.yml [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+      # Define the headers
+      headers=(
+        -H "Authorization: Basic $patEncoded"
+        -H "Content-Type: application/json"
+      )
+
+      # Get the pool ID
+      url="https://dev.azure.com/${organization}/_apis/distributedtask/pools?poolName=${poolName}&api-version=7.1"
```
[the remark this answers is not in the document]
Comment (josesimoes) [reply azure-pipelines-templates/device-bootstrap.yml:44]: same here

File: azure-pipelines-templates/device-bootstrap.yml [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+      AZP_POOL_AGENTS=$(curl -s "${headers[@]}" -X GET "$url")
+      poolId=$(echo "$AZP_POOL_AGENTS" | jq -r '.value[0].id')
+
+      echo "Pool ID: $poolId"
```
[the remark this answers is not in the document]
Comment (josesimoes) [reply azure-pipelines-templates/device-bootstrap.yml:48]: I suppose we're assuming that poolId will always be there, correct? (nothing against it)

File: azure-pipelines-templates/device-bootstrap.yml [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+              fi
+          else
+              echo "Failed to get pipeline run status. HTTP Status Code: $http_code"
+              echo "Response: $content"
+              exit 1
+          fi
+
+          # Wait for a while before checking again
+          sleep 30
+      done
```
[the remark this answers is not in the document]
Comment (josesimoes) [reply azure-pipelines-templates/device-bootstrap.yml:144]: I find the suggestion for the timeout valuable. Not so much for the retry mechanism, we don't need that.
I see value too in using a parameter for the timeout. Let's have it with a reasonable value which will work for most repos, but still have the change to set it (I'm thinking on IoT bindings runs that can take a couple of hours on some cases).

File: azure-pipelines-templates/device-test.yml [hunk 5]
```diff
@@ -207,6 +207,9 @@ stages:
                     elif [[ "$RESULT_TEST" =~ "The argument .* is invalid." ]]; then
                         echo "Test failed as arguments are not valid."
                         exit 1
+                    elif tail -n 5 <<< "$RESULT_TEST" | grep -qE "Failed: [1-9][0-9]*"; then
+                        echo "Test failed with non-zero failures."
+                        exit 1
```
[the remark this answers is not in the document]
Comment (josesimoes) [reply azure-pipelines-templates/device-test.yml:212]: This is on purpose. We only need this in Linux.

**Review by josesimoes — APPROVED** [review]
Looks great.
I've added my comments on coderabitai suggestion. Some are pretty valid and make sense to add. If not now, on another pass. Up to you.

File: azure-pipelines-templates/device-bootstrap.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (Ellerbach) [reply azure-pipelines-templates/device-bootstrap.yml:44]: I prefer to have the full URL as in the documentation, makes it easier to understand what is what

File: azure-pipelines-templates/device-bootstrap.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (Ellerbach) [reply azure-pipelines-templates/device-bootstrap.yml:48]: the script will fail anyway later on with a display null value

File: azure-pipelines-templates/device-test.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (Ellerbach) [reply azure-pipelines-templates/device-test.yml:212]: not needed for Windows! It just works out of the box

**Pull request merged**